### DISCUSSION TITLE
How to keep history data from previous solarman?

### DISCUSSION THREAD

**Discussion by Pigi-102** [davidrapan/ha-solarman#127] [Q&A] [answered]
I have a question: I have the old solarman integration, and quite a lot of historical  data that I'm interessed in, but the old integration does not seems to be maintained costantly, or at least seems that it's quite hard to get new feature and request implemented ( or even pull request merged ).
If I install this integration, would my data be keepd or I'm going to start all over again ?
As far as I can understand it should, but I would like to be sure, if possible.
Thanks in advance.

Pigi_102

**Comment by PV-Joe** [comment]
Look here how to merge history:
https://github.com/davidrapan/ha-solarman/wiki/Naming-Scheme

**Comment by Pigi-102** [comment]
Interesting !
So I should diff the new and old yaml for the inverter definitions to find the differences, and then apply them as you described ?
As I use a definition file made by myself ( basically the sofar_g3hyd.yaml modified to remove all stuff from 3Phase version, to make it single phase ), I could do the same thing, but keeping my names ( for the existing sensor names ) and this should makes a transparent passage .
Am I correct ?

**Comment by PV-Joe** [comment]
Check if there is a definition yaml for your device here:

https://github.com/davidrapan/ha-solarman/wiki/Supported-Devices

I think you can use your old yaml as well.
Put it in a new folder called custom in the same directory of the other definitions so it won‘t be deleted or overwritten by future updates.

Do a full backup before you start. Then you can test and if it doesn’t deserve your needs you can restore.

This integration is tons better than the old one. I‘m using it with a Deye inverter and it‘s 100% stable and fetches data every 5 seconds.

Good luck!

**Comment by Jan-Ole-Menz** [comment]
Hello,

I migrated today to this integration and my today's and total values are now tracked in a different entity, how to I merge them back together?

Example:

Old: wechselrichter_daily_energy_sold
New: wechselrichter_today_energy_export

**Reply by PV-Joe** [reply]
Does this help?

https://github.com/davidrapan/ha-solarman/wiki/Naming-Scheme

**Reply by Jan-Ole-Menz** [reply]
Sadly no, just get this error when trying:

![image](https://github.com/user-attachments/assets/03b8b71c-f1d6-489f-aaf5-0377afb4c41d)

**Reply by davidrapan** [reply]
You gonna have to remove the device and do it all through Developer tools > STATISTICS tab.

_And don't forget: First check if entity with that new name doesn't already exist and if does delete it using STATISTICS tab in Developer tools!_

**Reply by Jan-Ole-Menz** [reply]
yes but the delete function isn't available:

![image](https://github.com/user-attachments/assets/f7e235d5-9faf-4f4f-8bc7-0e048be1b0ab)

**Reply by davidrapan** [reply]
I mean delete whole device from integration, not just single sensor.

**Reply by davidrapan** [reply]
**And you will have to click on Fix Issue and delete data on all sensors WITH THE NEW NAME! Before transferring the history from old to new name. If you don't do that first no history will be transferred.**

**Reply by Jan-Ole-Menz** [reply]
ok.

But before I do that I have to create a mapping table for the 34 entities that now have a new name, correct?
And I will lose the data on the new entities entered in the last 4 hours, correct?

- the link on the wiki for this goes to nowhere
![image](https://github.com/user-attachments/assets/ffc289c1-4247-4ed7-8efa-1e5296d135f2)

Edit: HA says it will delete all entities if I delete the device, I guess there is another way?
![image](https://github.com/user-attachments/assets/98b6d3f4-5237-4756-9470-f622213b7b8e)

**Reply by davidrapan** [reply]
> But before I do that I have to create a mapping table for the 34 entities that now have a new name, correct?
And I will lose the data on the new entities entered in the last 4 hours, correct?

Yes and Yes. And the most important are total sensors as they are in the Energy dashboard, etc.

> the link on the wiki for this goes to nowhere

Thanks for pointing it out I'll fix it.

> Edit: HA says it will delete all entities if I delete the device, I guess there is another way?

That's the correct way. It will just delete their "physical" representation not the history.

**Reply by Jan-Ole-Menz** [reply]
Thanks, it seems I'm stuck again though.

I deleted the device, and 10 of the new entities. Then i wanted to set the new ID but cannot get there and get this error:

![image](https://github.com/user-attachments/assets/e35f0ce6-3b83-431b-b0b9-b64c152cbc7a)

**Reply by davidrapan** [reply]
Hmmm, there are many ways how to get around that. They simplest for you will be to just load up the previous integration. Change the entity_ids and then delete the integration and add back this one again.

**Reply by Jan-Ole-Menz** [reply]
oh ok, not that simple after all ;)

**Reply by davidrapan** [reply]
It depends. Due to the complexity of HA many different scenarios can happen.

_Best is if you catch it before the update and "prepare" for it._

Then it became progressively more difficult how far beyond that point you are.

To be honest if there would be a way how to avoid it all I would gladly choose that route but the state of inverter profiling which was adopted from previous integration was unmaintainable in the long run and it had to be changed rather sooner than later. Best would be, obviously, if that had been dealt with from the start but... you never know during the development.

And for rest of the cases there are custom profiles. 😉

**Reply by Jan-Ole-Menz** [reply]
Yeah I didn't account for any name changes, my bad.

I now did all of that and it created duplicates of the ?Migrated? ID's -->

![image](https://github.com/user-attachments/assets/2f45457d-e14f-4b73-8988-3bd0c6fd1794)

**Reply by davidrapan** [reply]
You have to explain the picture bit more as it is rather confusing.

**Reply by Jan-Ole-Menz** [reply]
First two lines are the old entities from the other solarman integration (old name on the left), migrated to the new naming (entity id, on the right)

The bottom row are automatically created after I reinstalled the new integration.

**Reply by davidrapan** [reply]
And did you remove the device before that?

**Reply by davidrapan** [reply]
It will append _# whenever the entity_id already exists.

**Reply by davidrapan** [reply]
And is also maybe wise to restart HA after every "remove" and "add" of integrations.

**Reply by Jan-Ole-Menz** [reply]
Yes I did.

First I deleted the device, then I uninstalled this Solarman integration (restart), then I reinstalled the old Solarman integration(restart), then migrated the entities, then installed the new integration(restart). 

Any idea of what I did wrong?

**Reply by davidrapan** [reply]
You should have also delete the device again before the last step (installing this integration)

**Reply by Jan-Ole-Menz** [reply]
oh :/

So what should I do now exactly?

**Reply by davidrapan** [reply]
It should be fixable just by deleting the device and re-adding it back again though.

**Reply by davidrapan** [reply]
But if you did not uninstalled the previous integration you could have some problems in the future so I would just delete the device and then uninstall integration and make sure that both are in the HACS as not installed.

And then proceed with installing this one.

**Reply by Jan-Ole-Menz** [reply]
It seems to work now, but somehow I also seem to have briked some data :/

![image](https://github.com/user-attachments/assets/4ac5b722-d920-40a2-ba31-bc9535288176)

**Reply by davidrapan** [reply]
**Firstly you are using wrong sensors in the energy dashboard. You should have used total sensors there.**

Secondly... you can adjust the data in the developer tools > STATISTICS tab by clicking on the little graph icon. Selecting problematic time point using datetime picker and putting zeroes instead those wrong values.

**Reply by Jan-Ole-Menz** [reply]
I didn't change my dashboard, and it looks perferctly fine outside of today 5 PM

Exemple of how yesterday looks right now:

![image](https://github.com/user-attachments/assets/265410f0-cf91-4efe-9f93-d0fc437dfdaf)

**Reply by Jan-Ole-Menz** [reply]
> Secondly... you can adjust the data in the developer tools > STATISTICS tab by clicking on the little graph icon. Selecting problematic time point using datetime picker and putting zeroes instead those wrong values.

will try that

**Reply by davidrapan** [reply]
Still.. somewhere in the HA docs is stated that sensors used in dashboard should be without any external resets. I think it's cause of the time zones and such so it is fully in control of the HA to decide when new day starts. 😉

And internal clocks of solarman are also highly unreliable. (Whole solarman is fking mess though...)

Edit: **Changing it from Today sensors to Total sensors should not cause you any history loss or anything. As it is tied to the history of the sensor itself.** 😉

**Reply by PV-Joe** [reply]
Your inverter is a Deye SUN 12K Hybrid right?

You have to choose the following definition file:

https://github.com/davidrapan/ha-solarman/blob/main/custom_components/solarman/inverter_definitions/deye_sg04lp3.yaml

It works perfect for my Deye SUN 8K Hybrid.

**Reply by Jan-Ole-Menz** [reply]
I now restored this morning's backup, as a lot of data seemed quite messed up, as well as automations and also messed up stuff after HA (influx DB, ..).

After restoring the backup I did the following: (gentle reminder for everyone to create a backup of their data before they migrate to this)


Extract current device configuration.

Relinked the following ID's: (probably not all) "daily" --> "today", ***_total_load_power --> ***_load_power, ***_total_grid_power --> ***_grid_power, "sold" --> "export", "bought" --> "import", "battery_soc" --> "battery"

Restarted HA

Deleted the solarman device

Restarted HA

Delete Solarman HACS integration

Restarted HA

Downloaded and installer this integration

Restarted HA

Added device

**Comment by philmassie** [comment, marked as answer]
Thanks so much for all the work on this integration @davidrapan . And also thanks for all your patience going over how to switch from the old to this new integration.
Reading over various posts I was still a little nervous about losing data, and how HA would handle history when a device got deleted.
I did lose a little bit of data during the process, in sort of unexpected ways. It seems like every entity that I renamed lost it's short term stats. But not entirely. For eg.
solarman battery, when you click the entity, there appears to be ~5 minutely data for the entire day. 
![Quickview](https://github.com/user-attachments/assets/1b9d00c2-d007-4504-9371-a8b44dd389f2)

When I click show more though, there is only short term data since the rename.
![History](https://github.com/user-attachments/assets/f3daa8fa-ca2c-4446-9f14-f9f7ddb6c16a)

This isn't that important, but its confusing to me how this works.


Nevertheless, I put together my overly-anxious-dummies guide, based on the above discussion. maybe it will help someone else. Use at your own risk of course and YMMMV.

Thanks again

1. go through the old and new {inverter_model}.yaml files comparing entity names
	- this is pretty tedious, VSCode compare files can help.
	- Isolate all the entities that have different names. In my case there were about 40 to change and VSCode looked like this:
![Clipboard01](https://github.com/user-attachments/assets/03ca0cd7-688d-470d-8008-0e404be60750)

	- the entity ID will be the name in lower case, with underscores instead of spaces and no with a 'solarman' prefix
	- eg "Daily Production" > "Today Production"
	- entity IDs were solarman_daily_production > solarman_today_production
2. Make a note of your inverter IP, port, serial number
	- Settings > Devices and services > Solarman Integration (old) > Configure
3. Make a backup
4. Rename all the entity IDs to match the new IDs
	- Settings > Devices and services > Solarman Integration (old) > 85 entities (yours may differ)
	- One entity at a time, search > select > gear wheel > change the Entity ID to match what the new one will be
	- Dont worry about the friendly names, the new integration seems to take care of these when you install it
5. Restart HA
6. Make a backup - last one
7. Delete the old Solarman device
	- Settings > Devices and services > Solarman Integration (old) > Devices > 3dots > Delete
	- Take a breath. My long term stats were still there, all the entities were gone from the UI though.
8. Restart HA
9. Delete old Solarman HACS integration
	- HACS > find the integration > 3dots > Remove
10. Restart HA
11. Install new solarman integration
12. Restart HA
13. Add your device
	- Settings > Devices and services > Add Integration > search for solarman
	- Wizard will prompt you for info about your inverter. Mine was prepropulated which was nice
	- Change the "Device name" to "solarman" so that all the entities get the same prefix as your old ones
	
That's it. I was relieved that everything went smoothly.

**Reply by davidrapan** [reply]
Hi @philmassie!

> For eg. solarman battery, when you click the entity, there appears to be ~5 minutely data for the entire day.

It's mentioned on the wiki: https://github.com/davidrapan/ha-solarman/wiki#-changes in point number 2. HA's recorder has default sampling rate of 5 minutes.

> Nevertheless, I put together my overly-anxious-dummies guide, based on the above discussion. maybe it will help someone else. Use at your own risk of course and YMMMV.

Awesome! I'll send anyone who will struggle with it to this post! Thanks for your very thorough step by step guide! 👍 

> Wizard will prompt you for info about your inverter. Mine was prepropulated which was nice

That's thanks to the solarman discovery build in the integration. 😉

**Comment by kefrens** [comment]
Thanks @philmassie  for the tutorial. Worked like a charm!

**Comment by deebeo-blip** [comment]
is renaming the entities still necessary? i noticed in the new config, there are always "alt" names for each entity.

e.g.
      - name: Total Energy Import
        alt: Total Energy Bought

do these alt names make renaming unneccessary?

**Reply by deebeo-blip** [reply]
nope, renaming seems necessary :(

**Reply by davidrapan** [reply]
Those alts are there just as a helpers to match those with old naming or names in documentation.

**Comment by raspitoaster** [comment]
I wanted to give the new solarman integration a try and found this "overly-anxious-dummies guide".
So I started to compare the device templates (**deye_2mppt.yaml**) of the old and the new integration and mapped the new ones to the old ones using the registers as reference. When renaming the entities, I found that two of them have the same name which is not accepted by Home Assistant.

![HASS solarman old vs new](https://github.com/user-attachments/assets/391b4deb-caa7-4c11-8063-65ee9d71271a)

Is this a bug, or do I miss something?
Both sensors would need to be renamed to "over_frequency_load_reduction" to replace the old integration, correct?



>" - name: "Over-Frequency Load Reduction"
        platform: number
        state_class: "measurement"
        uom: "%"
        rule: 1
        registers: [0x0023]
        configurable:
          min: 0
          max: 100
          mode: box
        range:
          min: 0
          max: 100
(...)
      - name: "Over-Frequency Load Reduction"
        platform: "switch"
        rule: 1
        registers: [0x0031]

**Reply by davidrapan** [reply]
They are under different platforms (number and switch) so it's totally fine.

**Reply by raspitoaster** [reply]
ok, I tried again on a cloned installation: same result.


![image](https://github.com/user-attachments/assets/9c435452-a1f2-4a25-811b-358a1c233d5d)



When trying to rename the ID of entity from the third line in the screenshot, Home Assistant does not accept it - the one with the number (the former "percentage") has already been renamed (first line in the screenshot) and (I assume) is causing the conflict. 

![image](https://github.com/user-attachments/assets/32c9612b-d4c2-4f68-9a64-d9290c4b0ad3)

**Comment by philmassie** [comment]
Hello. I'm not at my computer right now so I can't follow 100%, but I guess those new entities should be `number.` and `switch.`, not `sensor.`

I can only guess but in the old integration, they were both `sensor.` which explains why you can't give them the same name.

If the history of these two entities isn't too important to you you could maybe just ignore them? But if you want to keep their histories you might need to copy assign the old history to the new entity id. I don't know if that works across platforms though (sensor to switch or sensor to number).

Sorry, mine were all sensors I think...

**Reply by davidrapan** [reply]
Yes, history is platform agnostic. And I also think that history for those particular entities is really not important at all. 😉

**Reply by raspitoaster** [reply]
Yes - I also thought about keeping only the interesting ones and just ignoring (all) the other sensors.
So, instead of renaming all the (in my case) 39 entities, I will
1. rename the ones I want to keep to match the name in the new solarman integration
2. restart HA
3. delete the device from the old solarman integration
4. restart HA
5. remove the old solarman integration
6. restart HA
7. install the new solarman integration
8. restart HA
9. add the device to the new solarman integration
10. clean up the entities

**Reply by philmassie** [reply]
Cool, let us know how you get on please.
